invenio_query_parser/contrib/elasticsearch/walkers/dsl.py

```diff
@@ -0,0 +1,180 @@
+                                      OrOp, RangeOp, RegexValue,
+                                      SingleQuotedValue, Value, ValueQuery)
+from invenio_query_parser.visitor import make_visitor
+
+
+class ElasticSearchDSL(object):
+    """Implement visitor to create Elastic Search DSL."""
+
+    visitor = make_visitor()
+
+    def __init__(self, keyword_to_fields=None):
+        """Provide a dictinary mapping from keywords to Elastic field(s)."""
+        self.keyword_to_fields = keyword_to_fields or {None: ['_all']}
+
+    def get_fields_for_keyword(self, keyword, mode='a'):
+        """Convert keyword to fields."""
+        field = self.keyword_to_fields.get(keyword, keyword)
+        if isinstance(field, dict):
+            field = field[mode]
+        elif isinstance(field, (list, tuple)):
+            return field
+        return [field]
+
+    # pylint: disable=W0613,E0102
+
+    @visitor(AndOp)
+    def visit(self, node, left, right):
+        return {'bool': {'must': [left, right]}}
+
+    @visitor(OrOp)
+    def visit(self, node, left, right):
+        return {'bool': {'should': [left, right]}}
+
+    @visitor(NotOp)
+    def visit(self, node, op):
+        return {'bool': {'must_not': [op]}}
+
+    @visitor(KeywordOp)
+    def visit(self, node, left, right):
+        if callable(right):
+            return right(left)
+        raise RuntimeError("Not supported second level operation.")
+
+    @visitor(ValueQuery)
+    def visit(self, node, op):
+        return op(None)
+
+    @visitor(Keyword)
+    def visit(self, node):
+        return node.value
+
+    @visitor(Value)
+    def visit(self, node):
+        def query(keyword):
+            fields = self.get_fields_for_keyword(keyword, mode='a')
+            return {
+                'multi_match': {
+                    'query': node.value,
+                    'fields': fields,
+                }
+            }
+        return query
+
+    @visitor(SingleQuotedValue)
+    def visit(self, node):
+        def query(keyword):
+            fields = self.get_fields_for_keyword(keyword, mode='p')
+            return {
+                'multi_match': {
+                    'query': node.value,
+                    'type': 'phrase',
+                    'fields': fields,
+                }
+            }
+        return query
+
+    @visitor(DoubleQuotedValue)
+    def visit(self, node):
+        def query(keyword):
+            fields = self.get_fields_for_keyword(keyword, mode='e')
+            if len(fields) > 1:
+                return {"bool": {"should": [
+                    {"term": {field: str(node.value)}} for field in fields
+                ]}}
+            return {'term': {field: node.value for field in fields}}
+        return query
+
+    @visitor(RegexValue)
+    def visit(self, node):
+        def query(keyword):
+            fields = self.get_fields_for_keyword(keyword, mode='r')
+            if len(fields) > 1:
+                res = {"bool": {"should": []}}
```

Comment:
note to myself: refactor like in https://github.com/inveniosoftware/invenio-query-parser/pull/41/files#diff-2cc2bfe6afba2c08dc65923ec65e9ccfR111